addons/purchase/models/account_invoice.py

```diff
@@ -148,119 +149,312 @@ def write(self, vals):
+            # reference that matches with an OCR invoice. We replace the invoice lines with the purchase order lines.
             self._set_purchase_orders(matched_po_lines.order_id, force_write=True)
 
-        elif method == 'subset_total_match': # don't erase and add autocomplete
+        elif method == 'subset_total_match':
+            # A subset of the referenced purchase order lines matches the total amount of this invoice.
+            # We keep the invoice lines, but add all the lines from the partially matched purchase orders:
+            #   * "naively" matched purchase order lines keep their quantity
+            #   * unmatched purchase order lines are added with their quantity set to 0
             self._set_purchase_orders(matched_po_lines.order_id, force_write=False)
 
-            with self._get_edi_creation() as move_form: # logic for unmatched lines
+            with self._get_edi_creation() as move_form:
                 unmatched_lines = move_form.invoice_line_ids.filtered(
                     lambda l: l.purchase_line_id and l.purchase_line_id not in matched_po_lines)
                 for line in unmatched_lines:
-                    if prefer_purchase_line:
-                        line.quantity = 0
-                    else:
-                        line.unlink()
+                    line.quantity = 0
+
+        elif method == 'subset_match':
+            # A subset of the referenced purchase order lines matches a subset of the invoice lines.
+            # We add the purchase order lines, but adjust the quantity to the quantities in the invoice.
+            # The original invoice lines that correspond with a purchase order line are removed.
+            self._set_purchase_orders(matched_po_lines.order_id, force_write=False)
 
-                if not prefer_purchase_line:
-                    move_form.invoice_line_ids.filtered('purchase_line_id').quantity = 0
+            with self._get_edi_creation() as move_form:
+                unmatched_lines = move_form.invoice_line_ids.filtered(
+                    lambda l: l.purchase_line_id and l.purchase_line_id not in matched_po_lines)
+                for line in unmatched_lines:
+                    line.unlink()
+
+                # We remove the original matched invoice lines and apply their quantities and taxes to the matched
+                # purchase order lines.
+                for purchase_line_id, invoice_line in matched_inv_lines.items():
+                    inv_line = move_form.invoice_line_ids.filtered(
+                        lambda l: l in invoice_line)
+                    po_line = move_form.invoice_line_ids.filtered(
+                        lambda l: l.purchase_line_id and l.purchase_line_id.id == purchase_line_id)
+                    if po_line:
+                        po_line.quantity = invoice_line.quantity
+                        po_line.tax_ids = invoice_line.tax_ids
+                    inv_line and inv_line.unlink()
```

Comment:
same for all that stuff

Reply:
I tried to rewrite it using `Command` methods.